src/caches/abstractbeliefpropagationcache.jl

```diff
@@ -24,26 +25,48 @@ function data_graph_type(bpc::AbstractBeliefPropagationCache)
 end
 data_graph(bpc::AbstractBeliefPropagationCache) = data_graph(tensornetwork(bpc))
 
-function default_message_update(contract_list::Vector{ITensor}; normalize=true, kwargs...)
-  sequence = contraction_sequence(contract_list; alg="optimal")
-  updated_messages = contract(contract_list; sequence, kwargs...)
+function message_update_function(
+  alg::Algorithm"contract",
+  contract_list::Vector{ITensor};
+  normalize=alg.kwargs.normalize,
+  sequence_alg=alg.kwargs.sequence_alg,
+)
+  sequence = contraction_sequence(contract_list; alg=sequence_alg)
+  updated_messages = contract(contract_list; sequence)
   message_norm = norm(updated_messages)
   if normalize && !iszero(message_norm)
     updated_messages /= message_norm
   end
   return ITensor[updated_messages]
 end
 
+function message_update_function(
+  alg::Algorithm"contract_custom_device",
+  contract_list::Vector{ITensor};
+  normalize=alg.kwargs.normalize,
+  sequence_alg=alg.kwargs.sequence_alg,
+  custom_device_adapt=alg.kwargs.adapt,
+)
+  adapted_contract_list = custom_device_adapt.(contract_list)
+  updated_messages = message_update_function(
+    Algorithm("contract"), adapted_contract_list; normalize, sequence_alg
+  )
+  dtype = datatype(first(contract_list))
+  return ITensor[adapt(dtype, updated_message) for updated_message in updated_messages]
+end
+
 #TODO: Take `dot` without precontracting the messages to allow scaling to more complex messages
 function message_diff(message_a::Vector{ITensor}, message_b::Vector{ITensor})
   lhs, rhs = contract(message_a), contract(message_b)
   f = abs2(dot(lhs / norm(lhs), rhs / norm(rhs)))
   return 1 - f
 end
 
-default_message(elt, inds_e) = ITensor[denseblocks(delta(elt, i)) for i in inds_e]
+function default_message(datatype, elt, inds_e)
+  ITensor[adapt(datatype, denseblocks(delta(elt, i))) for i in inds_e]
+end
```

Comment:
I think it is confusing passing both the `datatype` and `eltype`, since normally the `eltype` will be specified as part of the `datatype`. Also I think there is a missing `return`, and I think we don't need to specify the type as `ITensor`.

Reply:
Note that we could have two versions:
```julia
function default_message(datatype::Type{<:AbstractArray}, inds_e)
  return [adapt(datatype, denseblocks(delta(i))) for i in inds_e]
end

function default_message(elt::Type{<:Number}, inds_e)
  return default_message(Vector{elt}, inds_e)
end
```